torch/fx/OVERVIEW.md

```diff
@@ -0,0 +1,121 @@
+
+Consider the following example:
+
+```python
+  class M(torch.nn.Module):
+      def forward(self, x):
+          return torch.relu(x)
+
+  m = M()
+  traced = symbolic_trace(m)
+```
+
+First, the parameter `x` is transformed into a Proxy object and the corresponding Node (a Node with op = “placeholder” and target = “x”) is added to the Graph. The next operation is `torch.relu`, which takes `x` as a parameter. Because `x` is a Proxy, `__torch_function__` is called on `torch.relu`. This, in turn, transforms `torch.relu` into a Proxy object and adds a `torch.relu` Node to the Graph.
+
+# The FX IR
+
+Symbolic tracing captures an intermediate representation (IR), which is represented as a doubly-linked list of Nodes.
+
+Node is the data structure that represents individual operations within a Graph. For the most part, Nodes represent callsites to various entities, such as operators, methods, and Modules (some exceptions include Nodes that specify function inputs and outputs). Each Node has a function specified by its `op` property. The Node semantics for each value of `op` are as follows:
+
+- `placeholder` represents a function input. The `name` attribute specifies the name this value will take on. `target` is similarly the name of the argument. `args` holds either: 1) nothing, or 2) a single argument denoting the default parameter of the function input. `kwargs` is don't-care. Placeholders correspond to the function parameters (e.g. `x`) in the graph printout.
+- `get_attr` retrieves a parameter from the module hierarchy. `name` is similarly the name the result of the fetch is assigned to. `target` is the fully-qualified name of the parameter's position in the module hierarchy. `args` and `kwargs` are don't-care
+- `call_function` applies a free function to some values. `name` is similarly the name of the value to assign to. `target` is the function to be applied. `args` and `kwargs` represent the arguments to the function, following the Python calling convention
+- `call_module` applies a module in the module hierarchy's `forward()` method to given arguments. `name` is as previous. `target` is the fully-qualified name of the module in the module hierarchy to call. `args` and `kwargs` represent the arguments to invoke the module on, *including the self argument*.
+- `call_method` calls a method on a value. `name` is as similar. `target` is the string name of the method to apply to the `self` argument. `args` and `kwargs` represent the arguments to invoke the module on, *including the self argument*
+- `output` contains the output of the traced function in its `args[0]` attribute. This corresponds to the "return" statement in the Graph printout.
+
+Each Node also has a reference to the Nodes that it takes as input (`input_nodes`) and the Nodes that use it (`users`). Although Nodes are represented as a doubly-linked list, the use-def relationships form an acyclic graph and can be traversed as such.
```

Comment:
Maybe something like "To facilitate easier analysis of data dependencies, `Node`s have read-only properties `input_nodes` and `users`, which specify which `Node`s in the graph are used by this node and use this node, respectively"

Also maybe we want to make `users` a read-only property :)